polkadot/xcm/pallet-xcm/src/benchmarking.rs

```diff
@@ -16,68 +16,157 @@
-use frame_support::weights::Weight;
+use frame_benchmarking::{benchmarks, whitelisted_caller, BenchmarkError, BenchmarkResult};
+use frame_support::{traits::Currency, weights::Weight};
 use frame_system::RawOrigin;
 use sp_std::prelude::*;
 use xcm::{latest::prelude::*, v2};
 
 type RuntimeOrigin<T> = <T as frame_system::Config>::RuntimeOrigin;
 
+// existential deposit multiplier
+const ED_MULTIPLIER: u32 = 100;
+
+/// Pallet we're benchmarking here.
+pub struct Pallet<T: Config>(crate::Pallet<T>);
+
+/// Trait that must be implemented by runtime to be able to benchmark pallet properly.
+pub trait Config: crate::Config {
+	/// A `MultiLocation` that can be reached via `XcmRouter`. Used only in benchmarks.
+	///
+	/// If `None`, the benchmarks that depend on a reachable destination will be skipped.
+	fn reachable_dest() -> Option<MultiLocation> {
+		None
+	}
+
+	/// A `(MultiAssets, MultiLocation)` pair representing assets and the destination they can
+	/// be teleported to. Used only in benchmarks.
+	///
+	/// Implementation should also make sure `dest` is reachable/connected.
+	///
+	/// If `None`, the benchmarks that depend on this will be skipped.
+	fn teleportable_assets_and_dest() -> Option<(MultiAssets, MultiLocation)> {
+		None
+	}
+
+	/// A `(MultiAssets, MultiLocation)` pair representing assets and the destination they can
+	/// be reserve-transferred to. Used only in benchmarks.
+	///
+	/// Implementation should also make sure `dest` is reachable/connected.
+	///
+	/// If `None`, the benchmarks that depend on this will be skipped.
+	fn reserve_transferable_assets_and_dest() -> Option<(MultiAssets, MultiLocation)> {
+		None
+	}
+}
+
 benchmarks! {
+	where_clause {
+		where
+			T: pallet_balances::Config,
+			<T as pallet_balances::Config>::Balance: From<u128> + Into<u128>,
+	}
 	send {
 		let send_origin =
 			T::SendXcmOrigin::try_successful_origin().map_err(|_| BenchmarkError::Weightless)?;
 		if T::SendXcmOrigin::try_origin(send_origin.clone()).is_err() {
 			return Err(BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX)))
 		}
 		let msg = Xcm(vec![ClearOrigin]);
-		let versioned_dest: VersionedMultiLocation = T::ReachableDest::get().ok_or(
+		let versioned_dest: VersionedMultiLocation = T::reachable_dest().ok_or(
 			BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX)),
 		)?
 		.into();
 		let versioned_msg = VersionedXcm::from(msg);
 	}: _<RuntimeOrigin<T>>(send_origin, Box::new(versioned_dest), Box::new(versioned_msg))
 
 	teleport_assets {
-		let asset: MultiAsset = (Here, 10).into();
-		let send_origin =
-			T::ExecuteXcmOrigin::try_successful_origin().map_err(|_| BenchmarkError::Weightless)?;
-		let origin_location = T::ExecuteXcmOrigin::try_origin(send_origin.clone())
+		let (assets, destination) = T::teleportable_assets_and_dest().ok_or(
+			BenchmarkError::Override(BenchmarkResult::from_weight(Weight::MAX)),
+		)?;
+
+		// most chains deploying `pallet-xcm` don't have `pallet-assets` so we're
```

Comment:
This doesn't seem like the right location to enforce this property. `teleportable_assets_and_dest` itself should instead enforce it. I also don't really find it probable that one would want to use more than 1 asset in the `teleport_assets` instruction benchmark -- it would almost seem nonsensical to me.

Reply:
yes, we could have each runtime implementation of `teleportable_assets_and_dest` do the required prefunding and everything, but will result in a lot of duplicated code and not really add much value to the benchmark

rather keep as is, but will change if you feel strongly about it

Reply:
What I mean by "enforcing" is simply have `teleportable_asset_and_dest` only return a single `MultiAsset`.

Reply:
ok, done